domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```diff
@@ -0,0 +1,470 @@
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().duration).isEqualTo(1000)
+  }
+
+  @Test
+  fun testObserver_preparePlayer_invokeChangeDataSource_capturesCorrectPosition() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.seekTo(500)
+    audioPlayerController.changeDataSource(TEST_URL2)
+    shadowMediaPlayer.invokePreparedListener()
+    audioPlayerController.play()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().position).isEqualTo(0)
+  }
+
+  @Test
+  fun testObserver_observeWithFragment_invokeRelease_checkRemoveObservers() {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+    playProgress.observe(fragment, mockAudioPlayerObserver)
+
+    audioPlayerController.play()
+    audioPlayerController.releaseMediaPlayer()
+
+    assertThat(shadowMediaPlayer.state).isEqualTo(ShadowMediaPlayer.State.END)
+    assertThat(playProgress.hasObservers()).isFalse()
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_preparePlayer_invokePauseAndAdvance_checkNoNewUpdates()
+      = runBlockingTest(coroutineContext) {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    audioPlayerController.pause()
+    advanceTimeBy(2000)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.PAUSED)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_preparePlayer_invokeCompletionAndAdvance_checkJobIsNull()
+      = runBlockingTest(coroutineContext) {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    shadowMediaPlayer.invokeCompletionListener()
+    advanceTimeBy(2000)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.COMPLETED)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_observeData_removeObserver_checkJobIsNull()
+      = runBlockingTest(coroutineContext) {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+    playProgress.observeForever(mockAudioPlayerObserver)
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    playProgress.removeObserver(mockAudioPlayerObserver)
+    advanceTimeBy(2000)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_addAndRemoveObservers_checkJobStates() = runBlockingTest(coroutineContext) {
```

Comment:
For naming: I think the expected outcome here is that the test doesn't hang, not ``checkJobStates``. Please update.

Ditto for other related tests above.